src/main/scala/mesosphere/marathon/core/task/Task.scala

```diff
@@ -63,17 +60,102 @@ import play.api.libs.json._
   * Marathon will notice spurious tasks in the offer and create the appropriate
   * unreserve operations. See https://github.com/mesosphere/marathon/issues/3223
   */
-sealed trait Task {
-  def taskId: Task.Id
-  def reservationWithVolumes: Option[Task.Reservation]
-  def runSpecVersion: Timestamp
 
-  /** apply the given operation to a task */
-  def update(update: TaskUpdateOperation): TaskUpdateEffect
+case class Task(taskId: Task.Id, runSpecVersion: Timestamp, status: Task.Status) {
+  import Task.log
 
   def runSpecId: PathId = taskId.runSpecId
 
-  def status: Task.Status
+  private[this] def hasStartedRunning: Boolean = status.startedAt.isDefined
+
+  /** apply the given operation to a task */
+  def update(instance: Instance, op: TaskUpdateOperation): TaskUpdateEffect = op match {
+
+    // exceptional case: the task is already terminal. Don't transition in this case.
+    // This might be because the task terminated (e.g. finished) before Marathon issues a kill Request
+    // to Mesos. Mesos will likely send back a TASK_LOST status update, because the task is no longer
+    // known in Mesos. We'll never want to transition from one terminal state to another as a terminal
+    // state should already be distinct enough.
+    // related to https://github.com/mesosphere/marathon/pull/4531
+
+    case op: TaskUpdateOperation if this.isTerminal =>
+      log.warn(s"received $op for terminal $taskId, ignoring")
+      TaskUpdateEffect.Noop
+
+    // case 1: running
+    case TaskUpdateOperation.MesosUpdate(Condition.Running, mesosStatus, now) if !hasStartedRunning =>
+      val updatedNetworkInfo = status.networkInfo.update(mesosStatus)
+      val updatedTask = copy(status = status.copy(
+        mesosStatus = Some(mesosStatus),
+        condition = Condition.Running,
+        startedAt = Some(now),
+        networkInfo = updatedNetworkInfo))
+      TaskUpdateEffect.Update(newState = updatedTask)
+
+    // case 2: terminal; extractor applies specific logic e.g. when an Unreachable task becomes Gone
+    case TaskUpdateOperation.MesosUpdate(newStatus: Terminal, mesosStatus, _) =>
+      val updatedStatus =
+        if (instance.hasReservation) {
+          status.copy(
+            mesosStatus = Some(mesosStatus),
+            // Note the task needs to transition to Reserved, otherwise the instance will not transition to Reserved
+            condition = Condition.Reserved)
+        } else {
+          status.copy(
+            mesosStatus = Some(mesosStatus),
+            condition = newStatus)
+        }
```

Comment:
Matter of taste. What do you think about

```scala
val newCondition = if (instance.hasReservation) Condition.Reserved else newStatus
val updatedStatus = status.copy(mesosStatus = Some(mesosStatus), condition = newCondition)

Reply:
Done. Thanks.